### PR TITLE
feat: Add more builtin evaluator forms and improve flattening utility

### PR DESCRIPTION
**Pull request by cephalization**
https://github.com/user-attachments/assets/64fd78ec-16a3-4000-bb2b-373d637b6599

- Added a new function to format evaluator display names in a consistent manner.
- Updated the CodeEvaluatorForm to include additional evaluator types: ExactMatch, Regex, LevenshteinDistance, and JSONDistance.
- Enhanced the flattenObject utility to support formatting of array indices and added relevant parameters for better control over output structure.
- Expanded test coverage for flattenObject to ensure comprehensive validation of new features and edge cases.

Depends on #10826

---

> [!NOTE]
> Introduces additional built-in evaluators end-to-end and improves input mapping utilities.
> 
> - Adds server-side evaluators: `ExactMatch`, `Regex`, `LevenshteinDistance`, `JSONDistance` (with helpers `levenshtein_distance` and `json_diff_count`) and enhances `Contains` with `require_all`
> - Implements matching UI forms and code samples for these evaluators and wires them into `CodeEvaluatorForm`
> - Adds `formatDisplayName` and uses it to derive default `displayName` when creating built-in evaluators
> - Enhances `flattenObject` with `formatIndices` (bracket index notation) and expands tests to cover new params and edge cases; updates usage in `EvaluatorInputMapping`
> - Updates tests to reflect new evaluator functionality and GraphQL evaluation result union shape (success/error) and adjusts mutation tests for duplicate display name constraints
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c389f66034bfda098d5b5d28a652b29693c9e68d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: app/src/components/evaluators/ExactMatchEvaluatorForm.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+              inputValue={expectedPath ?? ""}
+              css={css`
+                width: 100%;
+                min-width: 0 !important;
+                .px-combobox-container {
+                  min-width: 0 !important;
+                  input {
+                    min-width: 0 !important;
+                  }
+                }
```
Comment (cephalization) [inline app/src/components/evaluators/ExactMatchEvaluatorForm.tsx:86]: All of these are necessary until we fix combobox

#10094

File: app/src/components/dataset/CreateBuiltInDatasetEvaluatorSlideover.tsx [hunk 2]
```diff
@@ -69,6 +69,21 @@ export function CreateBuiltInDatasetEvaluatorSlideover({
...
+      .replace(/([a-z0-9])([A-Z])/g, "$1_$2") // Handle boundary between lower and upper: fooBar -> foo_Bar
+      // replace spaces with underscores
+      .replace(/\s+/g, "_")
+      // trim leading and trailing underscores
+      .replace(/^_+|_+$/g, "")
+      // convert to lowercase
+      .toLowerCase()
+  );
+};
+
```
Comment (axiomofjoy) [inline app/src/components/dataset/CreateBuiltInDatasetEvaluatorSlideover.tsx:86]: I think it would make sense to add a display name to the built-in evaluators types defined in the backend and GraphQL API.
